File: packages/ink-calligrapher/src/charmap.ts
/**
 * Character set of the calligrapher model, extracted verbatim from the
 * vendored engine's `H` map (vendor/calligrapher-ai/engine.pretty.js).
 * Ids 0-3 are control codes: 0 pad, 1 unknown, 2 start-of-text,
 * 3 end-of-text. Text is encoded as [START, ...chars, END], with
 * unmapped characters falling back to UNKNOWN.
 */

export const UNKNOWN = 1;
export const START = 2;
export const END = 3;

const ENTRIES: Array<[string, number]> = [
  ['"', 4], ["M", 5], ["r", 6], [".", 7], [" ", 8], ["A", 9], ["z", 10],
  ["u", 11], ["m", 12], ["i", 13], ["a", 14], ["n", 15], ["'", 16],
  ["s", 17], ["S", 18], ["e", 19], ["c", 20], ["t", 21], ["y", 22],
  ["I", 23], ["w", 24], ["o", 25], ["l", 26], ["d", 27], ["k", 28],
  ["p", 29], ["h", 30], ["T", 31], ["b", 32], ["g", 33], ["v", 34],
  ["f", 35], ["O", 36], [",", 37], ["N", 38], ["V", 39], ["-", 40],
  ["H", 41], ["E", 42], ["j", 43], ["x", 44], ["G", 45], ["P", 46],
  ["B", 47], ["L", 48], ["q", 49], ["Y", 50], ["?", 51], ["D", 52],
  ["F", 53], ["W", 54], ["R", 55], ["#", 56], ["C", 57], ["K", 58],
  ["1", 59], ["9", 60], ["5", 61], ["0", 62], ["2", 63], ["J", 64],
  ["U", 65], ["(", 66], [")", 67], ["4", 68], ["3", 69], ["7", 70],
  ["6", 71], ["!", 72], [";", 73], [":", 74], ["Q", 75], ["8", 76],
  ["/", 77], ["Z", 78], ["X", 79], ["*", 80], ["[", 81], ["+", 82],
  ["]", 83], ["&", 84],
];

export const CHAR_TO_ID: ReadonlyMap<string, number> = new Map(ENTRIES);

/** Every printable character the model can write. */
export const ALPHABET: readonly string[] = ENTRIES.map(([character]) => character);

File: packages/ink-kotlin/ink-calligrapher/src/main/kotlin/com/trylonghand/ink/calligrapher/Charmap.kt
/// Character set of the calligrapher model, ported verbatim from
/// packages/ink-calligrapher/src/charmap.ts (itself extracted from the
/// vendored engine's `H` map). Ids 0-3 are control codes: 0 pad, 1 unknown,
/// 2 start-of-text, 3 end-of-text. Text is encoded as [START, ...chars,
/// END], with unmapped characters falling back to UNKNOWN.

package com.trylonghand.ink.calligrapher

internal const val UNKNOWN: Int = 1
internal const val START: Int = 2
internal const val END: Int = 3

private val entries: List<Pair<Char, Int>> = listOf(
    '"' to 4, 'M' to 5, 'r' to 6, '.' to 7, ' ' to 8, 'A' to 9, 'z' to 10,
    'u' to 11, 'm' to 12, 'i' to 13, 'a' to 14, 'n' to 15, '\'' to 16,
    's' to 17, 'S' to 18, 'e' to 19, 'c' to 20, 't' to 21, 'y' to 22,
    'I' to 23, 'w' to 24, 'o' to 25, 'l' to 26, 'd' to 27, 'k' to 28,
    'p' to 29, 'h' to 30, 'T' to 31, 'b' to 32, 'g' to 33, 'v' to 34,
    'f' to 35, 'O' to 36, ',' to 37, 'N' to 38, 'V' to 39, '-' to 40,
    'H' to 41, 'E' to 42, 'j' to 43, 'x' to 44, 'G' to 45, 'P' to 46,
    'B' to 47, 'L' to 48, 'q' to 49, 'Y' to 50, '?' to 51, 'D' to 52,
    'F' to 53, 'W' to 54, 'R' to 55, '#' to 56, 'C' to 57, 'K' to 58,
    '1' to 59, '9' to 60, '5' to 61, '0' to 62, '2' to 63, 'J' to 64,
    'U' to 65, '(' to 66, ')' to 67, '4' to 68, '3' to 69, '7' to 70,
    '6' to 71, '!' to 72, ';' to 73, ':' to 74, 'Q' to 75, '8' to 76,
    '/' to 77, 'Z' to 78, 'X' to 79, '*' to 80, '[' to 81, '+' to 82,
    ']' to 83, '&' to 84,
)

internal val charToId: Map<Char, Int> = entries.toMap()

/** Every printable character the model can write. */
public val calligrapherAlphabet: List<Char> = entries.map { it.first }
